tsdb/exemplar.go

```diff
@@ -0,0 +1,209 @@
+		// when the exemplar from the last loop iteration was written has since been overwritten
+		// with an exemplar from another series.
+		// todo (callum) confirm if this check is still needed now that adding an exemplar should
+		// update the index and previous pointer for the series whose exemplar was overwritten.
+		if idx == -1 || string(ce.exemplars[idx].seriesLabels.Bytes(buf)) != string(l.Bytes(buf)) {
+			break
+		}
+
+		e = ce.exemplars[idx]
+		// todo (callum) This line is needed to avoid an infinite loop, consider redesign of buffer entry struct.
+		if ce.exemplars[idx].scrapeTimestamp > lastTs {
+			break
+		}
+
+		lastTs = ce.exemplars[idx].scrapeTimestamp
+		if e.scrapeTimestamp >= start && e.scrapeTimestamp <= end {
+			ret = append(ret, exemplar.ExemplarScrapeTimestamp{Exemplar: e.exemplar, ScrapeTimestamp: e.scrapeTimestamp})
+		}
+		idx = ce.exemplars[idx].prev
+	}
+	reverseExemplars(ret)
+	return ret, nil
+}
+
+// Takes the circularBufferEntry that will be overwritten and updates the
+// storages index for that entries labelset if necessary.
+func (ce *CircularExemplarStorage) indexGcCheck(cbe *circularBufferEntry) {
+	if cbe == nil {
+		return
+	}
+
+	l := cbe.seriesLabels
+	i := cbe.prev
+	if cbe.prev == -1 {
+		delete(ce.index, l.String())
+		return
+	}
+
+	l2 := ce.exemplars[i].seriesLabels
+	if !labels.Equal(l2, l) { // No more exemplars for series l.
+		delete(ce.index, cbe.seriesLabels.String())
+		return
+	}
+	// There's still at least one exemplar for the series l, so we can update the index.
+	ce.index[l.String()] = i
+}
+
+func (ce *CircularExemplarStorage) AddExemplar(l labels.Labels, t int64, e exemplar.Exemplar) error {
+	seriesLabels := l.String()
+	ce.lock.Lock()
+	defer ce.lock.Unlock()
+	idx, ok := ce.index[seriesLabels]
+
+	if !ok {
+		ce.indexGcCheck(ce.exemplars[ce.nextIndex])
+		// Default the prev value to -1 (which we use to detect that we've iterated through all exemplars for a series in Select)
+		// since this is the first exemplar stored for this series.
+		ce.exemplars[ce.nextIndex] = &circularBufferEntry{
+			exemplar:        e,
+			seriesLabels:    l,
+			scrapeTimestamp: t,
+			prev:            -1}
+		ce.index[seriesLabels] = ce.nextIndex
+		ce.nextIndex++
+		if ce.nextIndex >= cap(ce.exemplars) {
+			ce.nextIndex = 0
+		}
+		return nil
+	}
+
+	// Check for duplicate vs last stored exemplar for this series.
+	if ce.exemplars[idx].exemplar.EqualsWithoutTimestamp(e) {
+		ce.metrics.duplicateExemplars.Inc()
+		return storage.ErrDuplicateExemplar
+	}
+	if e.Ts <= ce.exemplars[idx].scrapeTimestamp || t <= ce.exemplars[idx].scrapeTimestamp {
+		ce.metrics.outOfOrderExemplars.Inc()
+		return storage.ErrOutOfOrderExemplar
+	}
+	ce.indexGcCheck(ce.exemplars[ce.nextIndex])
+	ce.exemplars[ce.nextIndex] = &circularBufferEntry{
+		exemplar:        e,
+		seriesLabels:    l,
+		scrapeTimestamp: t,
+		prev:            idx}
+	ce.index[seriesLabels] = ce.nextIndex
```

Comment:
Is the index the oldest or the newest exemplar? This code has it as the newest, the gc code as the oldest.

This whole logic doesn't seem right at all.